### PR TITLE
Demonstrating script address discovery and expose shared scripts in ApiWallet

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

https://jira.iohk.io/browse/ADP-525

# Overview

- [x] Extend ApiWallet by sharedScripts
- [x] Updating testing layer of Api.Types
- [x] introducing `Metadata` having `TxMetadata` and `Script` and update code base
- [x] update tests wrt the above change
- [x] enable sending script via auxliaryScript in TxBody
- [x] discover scripts from SL.Tx in fromAllegra
- [x] use isShared in Model
- [x] add integration tests showing that the script with our verification keys is later discovered and exposed in sharedScripts in ApiWallet
- [ ] add more integration test with to wallets and script containing both verification keys
- [ ] deal with db migration

Prerequisite for using isShared:
- [x] PR merged -> https://github.com/input-output-hk/cardano-wallet/pull/2308 

# Comments

### CODE REVIEW

File: specifications/api/swagger.yaml [hunk 1]
```diff
@@ -251,6 +251,13 @@ x-amount: &amount
       enum:
         - lovelace
 
+x-scriptStatus: &scriptStatus
+  type: string
+  enum:
+    - awaits_signing
+    - spent
+    - staked
```
Comment (KtorZ) [inline specifications/api/swagger.yaml]: I am not sure to understand the rationale behind those statuses? From our point of view, a script address exists on the blockchain or it doesn't. It'd be valuable however to show the balance available at that script address.

File: specifications/api/swagger.yaml [hunk 2]
```diff
@@ -1409,12 +1416,37 @@ components:
...
+    ApiSharedScript: &ApiSharedScript
+      description: Discovered script that contains our keys
+      type: object
+      required:
+        - script_address
+        - our_keys
+        - funds_status
+        - funds_available
+      properties:
+        script_address: *AnyAddress
```
Comment (KtorZ) [inline specifications/api/swagger.yaml]: Seems like there's some excessive wrapping here. The resulting JSON will look like `{ "script_address": { "address": "..." } }`

What about simply showing it as `"address": ...` ?

File: specifications/api/swagger.yaml [hunk 3]
```diff
@@ -1409,12 +1416,37 @@ components:
...
+        our_keys:
+          type: array
+          minItems: 1
+          items: *ApiVerificationKey
+        funds_status: *scriptStatus
+        funds_available: *amount
+
```
Comment (KtorZ) [inline specifications/api/swagger.yaml:1440]: I think it'll be valuable to show a JSON representation of the whole script no?

File: specifications/api/swagger.yaml (hunk 1 above)
Comment (paweljakubas) [reply specifications/api/swagger.yaml]: I expect to have the following lifecycle of addresses with script:
1. some funds are sent to this address
2. even more money can be sent
3. tx with script with all required signatures is sent
 and depending on where script is embedded, we have money spent or money staked
1-2 -> script awaits signing but available balance can be different
3 -> funds are unlocked and are spent or staked

File: lib/core/test/data/Cardano/Wallet/Api/ApiWallet.json [hunk 4]
```diff
@@ -1,476 +1,15262 @@
 {
-    "seed": 287889657345073705,
+    "seed": -7459132351643344646,
```
Comment (KtorZ) [inline lib/core/test/data/Cardano/Wallet/Api/ApiWallet.json]: That file is _way_ too large :grimacing:

**Comment by paweljakubas** [comment]
closing in favor of https://github.com/input-output-hk/cardano-wallet/pull/2675

**Pull request closed without merging**